tests/test_util.py

```diff
@@ -0,0 +1,25 @@
+from typing import Callable
+
+from inspect import getsource
+import ast
+
+from dusk.grammar import Grammar
+
+from dawn4py.serialization import make_sir
+from dawn4py.serialization.SIR import GridType
+from dawn4py._dawn4py import run_optimizer_sir
+
+
+def transpile_and_validate(stencil: Callable) -> None:
```

Comment:
From how this utility is defined, it seems that all the tests are letting some dusk code through the toolchain and expect no error to be thrown. In case of error, i guess this should be printed out and one should be able to understand what  went wrong (e.g. parsing, translation to sir ...).
However this kind of tests are very limited. One cannot test if the software gave the correct output when given a certain input, but only if an error has occurred (any error) while running. Also one might want to test if a particular error is (correctly) outputted when a flawed input is given.
Maybe we should revise how we want to test dusk. I would say that we should have unit tests taking dusk stencils as inputs and checking the SIR outputs against references. Maybe we should also have integration tests, involving the rest of the toolchain, but they should also check the correctness of the outputs somehow (we already have e2e tests, so it's debatable if we need integration tests).

Reply:
That's correct. The idea of this PR is basically, to start with a very _lightweight_ way of testing and then gradually improve it as we continue adding features. So it would develop like this:

1) Originally, _testing_ was to run `PYTHONPATH=./ dusk/cli.py test.py` manually (only _positive tests_)
2) Now, we can do `tox` or `pytest` and tests from separate files can be run, but still only _positive tests_ (also with github actions we get continuous testing).
3) In the future, we should have more test cases and also include _negative tests_ that check for correct exceptions.

However, what we want to test for _negative tests_ isn't quite clear. Originally, we said that dusk would do only the minimal checks and dawn's validator has to do the rest. This isn't really the case anymore, now it's a mix between both. And dusk will probably get more checks, especially when it implements semantics that doesn't exist in dawn.

I think next it would be good to check _compatibility_ with both ico backends and then we should probably think about what exceptions should be tested. Maybe, we should track this in a separate issue as well.

Reply:
Ok. Some comments to check if we are on the same page.
It's good practice to test, in the unit tests, only the functionality that is provided in the same module. As you say, there are two kinds of checks: those performed by dusk and those performed by dawn. 

- Via (negative) unit tests we should test checks performed by dusk

- Via (negative) integration tests (or e2e tests?) we should test checks performed by dawn (when starting from a dusk stencil)

So I see two tracks here for progressing with testing:

- Make these naive tests into full integration/e2e tests, i.e. categorize them as such and add checks on the output (compare IIR or generated code versus references). Also add (and keep distinct) some negative test cases

- Add positive and negative unit tests, checking only dusk functionalities (compare output against SIR references)

In any case it's ok to keep, for the moment, this PR's starting tests setup.

Reply:
We had a lengthy discussion on this. We agreed that we can keep it as is for this PR. It's not quite clear what approach should be chosen when the testing infrastructure will be improved.
Using serialized reference SIR might be necessary, since dawn isn't that suitable for validating SIR.